lib/interface_quda.cpp

```diff
@@ -3251,6 +3230,328 @@ void invertQuda(void *hp_x, void *hp_b, QudaInvertParam *param)
   profilerStop(__func__);
 }
 
+void loadFatLongGaugeQuda(QudaInvertParam *inv_param, QudaGaugeParam *gauge_param, void *milc_fatlinks, void *milc_longlinks)
+{
+  auto link_recon = gauge_param->reconstruct;
+  auto link_recon_sloppy = gauge_param->reconstruct_sloppy;
+  auto link_recon_precondition = gauge_param->reconstruct_precondition;
+
+  // Specific gauge parameters for MILC
+  int pad_size = 0;
+#ifdef MULTI_GPU
+  int x_face_size = gauge_param->X[1] * gauge_param->X[2] * gauge_param->X[3] / 2;
+  int y_face_size = gauge_param->X[0] * gauge_param->X[2] * gauge_param->X[3] / 2;
+  int z_face_size = gauge_param->X[0] * gauge_param->X[1] * gauge_param->X[3] / 2;
+  int t_face_size = gauge_param->X[0] * gauge_param->X[1] * gauge_param->X[2] / 2;
+  pad_size = MAX(x_face_size, y_face_size);
+  pad_size = MAX(pad_size, z_face_size);
+  pad_size = MAX(pad_size, t_face_size);
+#endif
+
+  int fat_pad = pad_size;
+  int link_pad = 3 * pad_size;
+
+  gauge_param->type = (inv_param->dslash_type == QUDA_STAGGERED_DSLASH || inv_param->dslash_type == QUDA_LAPLACE_DSLASH) ?
+    QUDA_SU3_LINKS :
+    QUDA_ASQTAD_FAT_LINKS;
+
+  gauge_param->ga_pad = fat_pad;
+  if (inv_param->dslash_type == QUDA_STAGGERED_DSLASH || inv_param->dslash_type == QUDA_LAPLACE_DSLASH) {
+    gauge_param->reconstruct = link_recon;
+    gauge_param->reconstruct_sloppy = link_recon_sloppy;
+    gauge_param->reconstruct_refinement_sloppy = link_recon_sloppy;
+  } else {
+    gauge_param->reconstruct = QUDA_RECONSTRUCT_NO;
+    gauge_param->reconstruct_sloppy = QUDA_RECONSTRUCT_NO;
+    gauge_param->reconstruct_refinement_sloppy = QUDA_RECONSTRUCT_NO;
+  }
+  gauge_param->reconstruct_precondition = QUDA_RECONSTRUCT_NO;
+
+  loadGaugeQuda(milc_fatlinks, gauge_param);
+
+  if (inv_param->dslash_type == QUDA_ASQTAD_DSLASH) {
+    gauge_param->type = QUDA_ASQTAD_LONG_LINKS;
+    gauge_param->ga_pad = link_pad;
+    gauge_param->staggered_phase_type = QUDA_STAGGERED_PHASE_NO;
+    gauge_param->reconstruct = link_recon;
+    gauge_param->reconstruct_sloppy = link_recon_sloppy;
+    gauge_param->reconstruct_refinement_sloppy = link_recon_sloppy;
+    gauge_param->reconstruct_precondition = link_recon_precondition;
+    loadGaugeQuda(milc_longlinks, gauge_param);
+  }
+}
+
+template <class Interface, class... Args>
+void callSplitGridQuda(Interface op, void **_hp_x, void **_hp_b, QudaInvertParam *param, void *h_gauge,
+                       void *milc_fatlinks, void *milc_longlinks, QudaGaugeParam *gauge_param, Args... args)
+{
+  /**
+    Here we first re-distribute gauge, color spinor, and clover field to sub-partitions, then call either invertQuda or dslashQuda.
+    - For clover field, we re-distribute cloverPrecise directly, after restore it after invertQuda/dslashQuda is called.
+    - For gauge field, we re-distribute the host side field, and restore it after.
+    - For color spinor field, we re-distribute the host side fields, and re-collect the host side fields.
+  */
+
```

Comment:
It would be good if we had a more flexible interface here, and allowed the user to specify whether gauge, clover and color spinor fields to be distributed are host or device.  E.g., give the option to use the `xxxPrecise` instances or use the host pointers.  It seems rather arbitrary to have support for fixed workflow of clover on device, and gauge/colorspinor on host.